lib/rubygems/package.rb

```diff
@@ -239,12 +239,14 @@ def add_metadata tar # :nodoc:
   ##
   # Builds this package based on the specification set by #spec=
 
-  def build skip_validation = false
+  def build skip_validation = false, allow_warnings = true
     Gem.load_yaml
     require 'rubygems/security'
 
     @spec.mark_version
-    @spec.validate unless skip_validation
+    unless skip_validation || !@spec.validate(true, allow_warnings)
+      raise Gem::InvalidSpecificationException, "Spec failed to validate"
```

Comment:
As a user of the `allow_warnings` option for Gem::Specification#validate I expect it to raise an exception when I set the flag for turning warnings into errors.

With your patch there are three results for an API user to handle:

1. No validation error, you can package the gem (but possible warning)
2. Validation error, you cannot package the gem, use `rescue` to inform their user.
3. Warning detected as an error, you cannot package the gem, you must now check the return value to know this.

If #validate returns `false` when an API user calls it directly (not through Gem::Package.build) with `allow_warnings` set to false then they have to reimplement raising an error or special case their own code in order to display the error correctly.

In order to discover this behavior they must read carefully, which is something users don't do.

By making #validate uniform in its handling of invalid portions of the specification users no longer have to read which reduces maintenance burden and makes use of the option's implementation and use much cleaner.  For example, we no longer need expressions with double-negatives like:

`unless skip_validation || !@spec.validate(true, allow_warnings)`